websocket-tcp-proxy/src/main.rs

```diff
@@ -105,6 +148,152 @@ async fn handle_socket(mut socket: WebSocket, ip: String, port: String) {
+    Path(identifier): Path<String>,
+    State(state): State<WSMap>,
+    ws: WebSocketUpgrade,
+    user_agent: Option<TypedHeader<headers::UserAgent>>,
+) -> impl IntoResponse {
+    tracing::info!("new mutiny websocket handler: {identifier}");
+    if let Some(TypedHeader(user_agent)) = user_agent {
+        tracing::info!("`{}` connected", user_agent.as_str());
+    }
+
+    ws.protocols(["binary"])
+        .on_upgrade(move |socket| handle_mutiny_ws(socket, identifier, params, state))
+}
+
+#[derive(Debug)]
+enum MutinyWSCommand {
+    Send { id: Bytes, val: Bytes },
+}
+
+/// handle_mutiny_ws will handle mutiny to mutiny (ws to ws) logic.
+/// A node pubkey will have a connection URL like: /v1/mutiny/{identifier}
+/// where identifier is either going to be arbitrary or based on their node
+/// pubkey. Future iterations might want a single identifier for all their
+/// nodes. This should be persistent enough to allow others to reconnect.
+///
+/// Owners:
+/// Need to send a signed message with the private key in order to
+/// verify that they are the owners. Afterwards, they will receive
+/// all incoming messages. If owner is already registered, kill new one.
+///
+/// Sending:
+/// Indicate which identifier you would like to message and the bytes to
+/// send. This should be the first 33 bytes of the message. If the owner
+/// is not connected or disconnections, the connection should be killed.
+/// This proxy will replace the 33 bytes with the identifier of the sender.
+///
+/// Receiving:
+/// You will receive a message with the first 33 bytes being the identifier
+/// that has sent the message and the rest of the bytes being the message.
+/// When replying to a received message, set the first 33 bytes to be the
+/// destination that had sent to you. IE, keeping same first 33 bytes.
+async fn handle_mutiny_ws(
+    mut socket: WebSocket,
+    identifier: String,
+    _params: Query<MutinyConnectionParams>,
+    state: WSMap,
+) {
+    // TODO do verification on the params and identifier
+    // This is important so that only the node with the
+    // private key can read and send messages through
+    // this socket.
+    let owner_id_bytes: Vec<u8> = if let Ok(b) = FromHex::from_hex(identifier.as_str()) {
+        b
+    } else {
+        // drop the connection if we didn't get a good pubkey hex
+        tracing::error!("could not parse hex string identifier");
+        return;
+    };
+    let owner_id_bytes = bytes::Bytes::from(owner_id_bytes);
+
+    // Now create one consumer and a producer that other
+    // mutiny websocket connections can reference to send
+    // to later. The consumer here is to listen to events
+    // that should be sent down the websocket that owns this.
+    let (tx, mut rx) = mpsc::channel::<MutinyWSCommand>(32);
+    state
+        .lock()
+        .await
+        .insert(owner_id_bytes.clone(), tx.clone());
+
+    tracing::debug!("listening for {identifier} websocket or consumer channel",);
+    loop {
```

Comment:
there's a bunch of TODOs in this loop, idk how needed those are now, might be good to bang out any of the easy ones